internal/cloud/state.go

```diff
@@ -81,6 +81,20 @@ of authorization and therefore this error can usually be fixed by upgrading the
 remote state version.
 `))
 
+const errUploadedStateNotFinalized = `State was uploaded to Terraform Cloud but not acknowledged by the platform, possibly due to an unexpected outage. State may be recovered either from the state-versions API or from the file system dump.
+
+The Terraform Cloud workspace will remain locked to allow for recovery. If running "terraform state push" to recover, add the "-lock=false" flag to instruct terraform not to first lock the workspace.
+
+After pushing the recovered state to Terraform Cloud, the workspace should then be unlocked:
+    terraform force-unlock %q`
+
+const backoffFinalizedMin = 1000
+const backoffFinalizedMax = 5000
+
+// StateFinalizedTimeout is the timeout duration for a given state upload to be acknowledged
+// as finalized by the Terraform Cloud API. See the associated error message above for details.
+var StateFinalizedTimeout = 5 * time.Minute
```

Comment:
One tricky thing about this interval and this change in general: state uploads are blocking, so Terraform won't start uploading a new state snapshot until the previous one is finished. In the case of a general platform slowdown, that means we may end up with pileups. Say an apply runs for three minutes and creates four state snapshots to be persisted - thanks to Terraform's graceful shutdown logic, the operation won't exit until either all four snapshots are finalized or 4 * 5 minutes pass.

This would be a fairly rare edge case, but so is the eventuality that this commit guards against! One way to proceed, depending on what we think the worst-case latency for finalizing a state version will be, might be to reduce this timeout so it's closer to the snapshot persist interval.